chapter_01/final_exercise_01.md

```diff
@@ -20,27 +19,177 @@
+ - [x] io.file.buffer.size - sets the buffer size for seqeunce files during read and write opperations (sequence files are binary files containing key-value paris which were created to solve some of the small file problems in hadoop by having multiple files inside one block file)
 
 ### Chapter 2: Hadoop Distributed File System (HDFS)
 
 6. **Q:** What is HDFS?
+6. **A:** HDFS is the Hadoop distributed system which is basically like a regular file system but it is distributed between multiple computers.
+HDFS is built for fault tolorence meaning, it was built with the idea that hardware failures are a norm and happen a lot so hdfs uses replication and redundency to keep the system highly available and fault tolorent.
+HDFS is the primary storage of the Hadoop applications, based on java and which contains data nodes and name nodes, the *name nodes* are the nodes which monitors all the data stored, replica counts and the amount of data nodes which are alive and the FS structure.
+the *data node* is the node that stores the data itself in the form of blocks which have a max size of a preconfigured amount, default is usually 128 MB but sometimes can be 64 or 256 MB by default.
+It is a rack aware system which means that it knows the physical topology of the network and the connected nodes which it uses to enhence fault tolorence by splitting replicas to different racks (an example with 3 replicas and 2 racks, hadoop will store one replica on the same rack with the name node and 2 more replicas on different racks in a different location / switch).
+
+provides High availability using 2 Name nodes with one being active and one being a standby, in case the active node will fail. (if the High availability mode is on)
+
+Hdfs uses also prefers reads from neighbours 'closer' to it (which are on the same rack) to minimize traffic between racks, meaning if someone needs to read a file, it reads the blocks from nodes which are on the same rack.
+
+Hdfs also has journal nodes which store changes to the file system (file addition, deletion and more) which is an edit log which is used to build the FSImage and the file system structure in the ram.
+
+in this mode, all of the datanodes send heartbeats and block reports to all namenodes to allow for faster failover.
+
+The journal nodes are there so that the namenode writes the edit log to the journal nodes as well, so that in the case of failure and in the case that the standby namenode is not in sync, the standby namenode will sync using the Editlogs from the journal nodes.
+
+The Journal nodes allow for reconstruction of the file system even in the case of disk falure of the Active name node and the Standby namenode by saving the Edit logs which can be used to reconstruct the FSImage in the case of hardware failure or losing even both the namenodes, using a cluster of journal nodes which all of the logs are written to all of them to allow for FSImage reconstruction even in the case a journal node fails.
```

Comment:
I still don't understand what is the role of the journal nodes? why do I need an odd number of them? why don't I just write the edit logs to a different source?